src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs

```diff
@@ -175,22 +165,287 @@ private async Task<long> DownloadBlocks(PeerInfo? bestPeer, BlocksRequest blocks
                     }
                 }
 
+                // Should only happen because of a lot of reorg
+                // or if the `HeaderLookupSize` become smaller significantly.
+                if (_downloadRequests.Count > headers.Count * 2)
+                {
+                    PruneRequestMap(headers);
+                }
+
                 if (blocksSynced > 0)
                 {
                     _syncReport.FullSyncBlocksDownloaded.Update(_blockTree.BestSuggestedHeader?.Number ?? 0);
                 }
-                else
+
+                _syncReport.FullSyncBlocksDownloaded.CurrentQueued = _downloadRequests.Count;
+
+                if (satisfiedEntry.Count == 0) // Nothing left to process
+                {
+                    return await AssembleRequest(headers, downloadReceipts, cancellation);
+                }
+            }
+        }
+
+        public void PruneDownloadBuffer()
+        {
+            _downloadRequests.Clear();
+        }
+
+        private void PruneRequestMap(IOwnedReadOnlyList<BlockHeader> currentHeaders)
+        {
+            HashSet<Hash256> currentHeaderHashes = currentHeaders.Select(h => h.Hash).ToHashSet();
+            HashSet<Hash256> toRemove = new HashSet<Hash256>();
```

Comment:
You don't need a set here.
```suggestion
            using ArrayPoolList<Hash256> toRemove = new ArrayPoolList<Hash256>();
```